File: components/users/src/com/github/joerodriguez/sbng2ex/UsersRepository.java
package com.github.joerodriguez.sbng2ex;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.jdbc.core.RowMapper;
import org.springframework.jdbc.core.namedparam.MapSqlParameterSource;
import org.springframework.jdbc.core.simple.SimpleJdbcInsert;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.Optional;

@Component
public class UsersRepository {

    private final JdbcTemplate jdbcTemplate;
    private final PasswordEncoder passwordEncoder;
    RowMapper<User> USER_ROW_MAPPER = (rs, rowNum) -> new User(rs.getLong("id"), rs.getString("email"));

    @Autowired
    public UsersRepository(JdbcTemplate jdbcTemplate, PasswordEncoder passwordEncoder) {
        this.jdbcTemplate = jdbcTemplate;
        this.passwordEncoder = passwordEncoder;
    }

    public User create(String email, String password) {
        MapSqlParameterSource params = new MapSqlParameterSource();
        params.addValue("email", email);
        params.addValue("password", passwordEncoder.encode(password));

        SimpleJdbcInsert simpleJdbcInsert = new SimpleJdbcInsert(jdbcTemplate)
                .withTableName("users")
                .usingGeneratedKeyColumns("id");

        Number generatedId = simpleJdbcInsert.executeAndReturnKey(params);

        return find(generatedId.longValue());
    }

    public Optional<User> findByEmail(String email) {
        List<User> users = jdbcTemplate.query(
                "SELECT id, email FROM users WHERE email = ?",
                new Object[]{email},
                USER_ROW_MAPPER
        );

        return Optional.ofNullable(users.isEmpty() ? null : users.get(0));
    }

    private User find(long id) {
        return jdbcTemplate.queryForObject(
                "SELECT id, email FROM users WHERE id = ?",
                new Object[]{id},
                USER_ROW_MAPPER
        );
    }

}

File: components/users/src/com/github/joerodriguez/sbng2ex/users/UsersRepository.kt
package com.github.joerodriguez.sbng2ex.users

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate
import org.springframework.jdbc.core.simple.SimpleJdbcInsert
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Component
import java.sql.ResultSet
import javax.sql.DataSource

@Component
class UsersRepository

    @Autowired
    constructor(
            private val dataSource: DataSource,
            private val jdbcTemplate: NamedParameterJdbcTemplate,
            private val passwordEncoder: PasswordEncoder
    )

{

    private val rowMapper: (ResultSet, Int) -> User = { rs, rowNum -> User(rs.getLong("id"), rs.getString("email")) }

    fun create(email: String, password: String): User {

        val params = mapOf(
                "email" to email,
                "password" to passwordEncoder.encode(password)
        )

        val insertedUserId = SimpleJdbcInsert(dataSource)
                .withTableName("users")
                .usingGeneratedKeyColumns("id")
                .executeAndReturnKey(params)

        return find(insertedUserId.toLong())
    }

    fun findByEmail(email: String): User? {

        val users = jdbcTemplate.query(
                "SELECT id, email FROM users WHERE email = :email",
                hashMapOf("email" to email),
                rowMapper
        )

        return if (users.isEmpty()) null else users[0]
    }

    private fun find(id: Long): User {

        return jdbcTemplate.queryForObject(
                "SELECT id, email FROM users WHERE id = :id",
                hashMapOf("id" to id),
                rowMapper
        )
    }

}
